saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/serviceprovider/authentication/OpenSamlAuthenticationManager.java

```diff
@@ -0,0 +1,340 @@
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import static java.lang.String.format;
+import static java.util.Collections.singleton;
+import static java.util.Collections.singletonList;
+import static org.springframework.util.Assert.notNull;
+import static org.springframework.util.StringUtils.hasText;
+
+/**
+ * @since 5.2
+ */
+public final class OpenSamlAuthenticationManager implements AuthenticationManager {
+
+	private static Log logger = LogFactory.getLog(OpenSamlAuthenticationManager.class);
+
+	private final OpenSamlImplementation saml = OpenSamlImplementation.getInstance();
+	private GrantedAuthoritiesMapper authoritiesMapper = (a -> a);
+	private Duration responseTimeValidationSkew = Duration.ofMillis(1000 * 60 * 5); // 5 minutes
+
+	private String getUsername(Saml2AuthenticationToken token, Assertion assertion) {
+		final Subject subject = assertion.getSubject();
+		if (subject == null) {
+			return null;
+		}
+		if (subject.getNameID() != null) {
+			return subject.getNameID().getValue();
+		}
+		if (subject.getEncryptedID() != null) {
+			NameID nameId = decrypt(token, subject.getEncryptedID());
+			return nameId.getValue();
+		}
+		return null;
+	}
+
+	private Assertion validateSaml2Response(Saml2AuthenticationToken token,
+											String recipient,
+											Response samlResponse) throws AuthenticationException {
+		if (hasText(samlResponse.getDestination()) && !recipient.equals(samlResponse.getDestination())) {
+			throw new Saml2Exception("Invalid SAML response destination: " + samlResponse.getDestination());
+		}
+
+		final String issuer = samlResponse.getIssuer().getValue();
+		if (logger.isDebugEnabled()) {
+			logger.debug("Processing SAML response from " + issuer);
+		}
+		if (token == null) {
+			throw new Saml2Exception(format("SAML 2 Provider for %s was not found.", issuer));
+		}
+		boolean responseSigned = hasValidSignature(samlResponse, token);
+		for (Assertion a : samlResponse.getAssertions()) {
+			if (isValidAssertion(recipient, a, token, !responseSigned)) {
+				return a;
+			}
+		}
+		for (EncryptedAssertion ea : samlResponse.getEncryptedAssertions()) {
+			Assertion a = decrypt(token, ea);
+			if (isValidAssertion(recipient, a, token, false)) {
+				return a;
+			}
+		}
+		throw new InsufficientAuthenticationException("Unable to find a valid assertion");
+	}
+
+	private boolean hasValidSignature(SignableSAMLObject samlResponse, Saml2AuthenticationToken token) {
+		if (!samlResponse.isSigned()) {
+			return false;
+		}
+
+		final List<X509Certificate> verificationKeys = getVerificationKeys(token);
+		if (verificationKeys.isEmpty()) {
+			return false;
+		}
+
+		for (X509Certificate key : verificationKeys) {
+			final Credential credential = getVerificationCredential(key);
+			try {
+				SignatureValidator.validate(samlResponse.getSignature(), credential);
+				return true;
+			}
+			catch (SignatureException ignored) {
+			}
```

Comment:
in my experience common programming technique for code hotspot. imagine if tomcat created objects and logged exceptions each time an incorrectly formatted http request came in. not all errors need to be logged or propagated. 

I'm inclined to do nothing here. But happy to put in better debug logging since that can be turned on and off at the app configuration level